pkg/utils/nodeclaim/nodeclaim.go

```diff
@@ -20,71 +20,142 @@ import (
 	"context"
 	"errors"
 	"fmt"
+	"strings"
 
 	"github.com/awslabs/operatorpkg/object"
+	"github.com/awslabs/operatorpkg/option"
+	"github.com/awslabs/operatorpkg/status"
 	"github.com/samber/lo"
 	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/types"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 
 	v1 "sigs.k8s.io/karpenter/pkg/apis/v1"
+	"sigs.k8s.io/karpenter/pkg/cloudprovider"
 )
 
+func IsManaged(nodeClaim *v1.NodeClaim, cp cloudprovider.CloudProvider) bool {
+	return lo.ContainsBy(cp.GetSupportedNodeClasses(), func(nodeClass status.Object) bool {
+		return object.GVK(nodeClass).GroupKind() == nodeClaim.Spec.NodeClassRef.GroupKind()
+	})
+}
+
+// IsManagedPredicates is used to filter controller-runtime NodeClaim watches to NodeClaims managed by the given cloudprovider.
+func IsManagedPredicates(cp cloudprovider.CloudProvider) predicate.Funcs {
+	return predicate.NewPredicateFuncs(func(o client.Object) bool {
+		return IsManaged(o.(*v1.NodeClaim), cp)
+	})
+}
+
+type ListOptions struct {
+	filterPredicates []func(*v1.NodeClaim) bool
+	listOptions      []client.ListOption
+}
+
+func resolveListOptions(opts ...option.Function[ListOptions]) *ListOptions {
+	resolved := option.Resolve(opts...)
+	if len(resolved.filterPredicates) == 0 {
+		resolved.filterPredicates = []func(*v1.NodeClaim) bool{func(nc *v1.NodeClaim) bool { return true }}
+	}
+	return resolved
+}
+
+func withClientListOptions(listOpts ...client.ListOption) option.Function[ListOptions] {
+	return func(opts *ListOptions) {
+		opts.listOptions = append(opts.listOptions, listOpts...)
+	}
+}
+
+func WithProviderIDFilter(providerID string) option.Function[ListOptions] {
+	return withClientListOptions(client.MatchingFields{"status.providerID": providerID})
+}
+
+func WithNodePoolFilter(nodePoolName string) option.Function[ListOptions] {
```

Comment:
```suggestion
func ForNodePool(nodePoolName string) option.Function[ListOptions] {
```

I wonder if this is more natural naming?